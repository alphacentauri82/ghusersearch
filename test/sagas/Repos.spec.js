import {
  put,
} from 'redux-saga/effects';
import {
  getRepos,
} from 'store/sagas/Repos';

describe('Saga: getRepos', () => {

  describe('when no cached data exists', () => {
    it('should call the API with the url and parameters', () => {
      const generator = getRepos({payload: {url: 'repos.net'}});
      generator.next();
      const repos = generator.next().value;
      expect(repos.CALL).toMatchSnapshot();
    });

    it('should put a success action with the normalized data', () => {
      const generator = getRepos({payload: {url: 'repos.net'}});
      const response = {
        data: [
          {id: 1, name: 'test'},
          {id: 2, name: 'hello'},
        ],
      };
      generator.next();
      generator.next();
      const normalized = generator.next(response).value;
      expect(normalized.PUT.action).toMatchSnapshot();
    });
  });

  describe('when cached data does exist', () => {
    it('should pass the result to the success action and not call the API', () => {
      const generator = getRepos({payload: {url: 'repos.net'}});
      generator.next();
      const successAction = generator.next(
        {
          result: [1, 2],
          entities: {},
        }
      ).value;

      expect(successAction).toEqual(
        put({
          meta: {fromCache: true},
          payload: {
            entities: {},
            result: [1, 2],
            url: 'repos.net',
          },
          type: 'REPOS_SUCCESS',
        })
      );
    });
  });

  describe('when a request to the API fails', () => {
    it('should call a failure action', () => {
      const generator = getRepos({payload: {url: 'repos.net'}});
      const apiError = new Error('it went wrong');
      apiError.response = 'test';

      generator.next();
      generator.next();
      const errorAction = generator.throw(apiError).value;
      expect(errorAction.PUT.action).toMatchSnapshot();
    });
  });

});
