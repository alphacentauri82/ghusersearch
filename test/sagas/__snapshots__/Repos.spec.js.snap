// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga: getRepos when a request to the API fails should call a failure action 1`] = `
Object {
  "error": true,
  "payload": [Error: it went wrong],
  "type": "REPOS_FAILURE",
}
`;

exports[`Saga: getRepos when no cached data exists should call the API with the url and parameters 1`] = `
Object {
  "args": Array [
    "repos.net",
    Object {
      "params": Object {
        "per_page": 8,
        "sort": "pushed",
      },
    },
  ],
  "context": null,
  "fn": [Function],
}
`;

exports[`Saga: getRepos when no cached data exists should put a success action with the normalized data 1`] = `
Object {
  "meta": Object {
    "fromCache": false,
  },
  "payload": Object {
    "entities": Object {
      "repos": Object {
        "1": Object {
          "name": "test",
        },
        "2": Object {
          "name": "hello",
        },
      },
    },
    "result": Array [
      1,
      2,
    ],
    "url": "repos.net",
  },
  "type": "REPOS_SUCCESS",
}
`;
